### ISSUE TITLE
Velero restore with restic not working for statefulsets with (multiple) initContainers

### ISSUE THREAD

**Issue by sgyuris** [vmware-tanzu/velero#3130]
[carlisia] Update on 1/13: this became a task.

-----
**What steps did you take and what happened:**
I have the latest azure AKS cluster 1.17.9 and 1.18.10 and azurefile as pv.

In the source cluster I run:
`velero backup create full-backup2 --default-volumes-to-restic --ttl "48h00m00s"`

The restic and velero backup part is done correctly. I can see it in the backup storage account. Everything is fine. The PV is backed up.

In the destination cluster (this should be a migration from one cluster to another) I run:
`velero restore create --from-backup full-backup2 --include-namespaces test-ns`

There are two statefulset in the very same test-ns namespace lets call it sts-test1 and sts-test2.

sts-test2 has no initcontainer.
sts-test1 need to wait for sts-test2 to be live. So we did introduced an initcontainer to check sts-test2's state and wait until sts-test2 is in working state. Then there is an another initcontainer in sts-test1 which do some copying and initial setup before the main container could start.

The thing is that sts-test2 (no initcontainer) got it's data back from the backup.
[time="2020-12-04T10:12:38Z" level=info msg="Restic backups for pod found" cmd=/velero logSource="pkg/restore/restic_restore_action.go:97" pluginName=velero pod=test-ns/sts-test2 restore=velero/full-backup-20201204101233]

sts-test1 has no log entry like you can see above and there is no backup restored. (backup is present in the blob store just like for sts-test2) but I see this in restic's log:
[time="2020-12-04T10:13:51Z" level=debug msg="Pod is not running restic-wait init container, not enqueuing restores for pod" controller=pod-volume-restore key=test-ns/sts-test1 logSource="pkg/controller/pod_volume_restore_controller.go:173"]

multiple times.

**What did you expect to happen:**

I expect that the pv content would be restored also for sts-test1.

**Environment:**

- Velero version (use `velero version`): 
Client:
        Version: v1.5.2
        Git commit: e115e5a191b1fdb5d379b62a35916115e77124a4
Server:
        Version: v1.5.2 
- Velero features (use `velero client config get features`):
 features: <NOT SET>
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
SRC cluster:
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9", GitCommit:"e3808385c7b3a3b86db714d67bdd266dc2b6ab62", GitTreeState:"clean", BuildDate:"2020-07-15T20:50:36Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
DEST cluster:
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.10", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-10-16T20:43:34Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
- Kubernetes installer & version:
AKS 1.17.9 and 1.18.10
- Cloud provider or hardware configuration:
MS Azure
- OS (e.g. from `/etc/os-release`):
Ubuntu 16.04

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by nrb** [comment]
@sgyuris Thanks for this issue! It's definitely a problem with our restic implementation that the init container is bound to the pod and doesn't bubble up to the statefulset, which results in the statefulset controller killing the pods in order to force the templates to match.

I thought we already had an issue for this, but I can't find one that represents it as well as what you've described, so I'm going to use this one and mark it for v1.7.0.

**Comment by bspa10** [comment]
Is there any workaround for this issue?

**Comment by reasonerjt** [comment]
@Lyndon-Li please investigate as this mechanism is also to be used by kopia.

**Comment by sseago** [comment]
There's a similar issue with restic restore and DeploymentConfigs in OpenShift. I believe the culprit with DCs is that they use ReplicationControllers instead of ReplicaSets, and upon restore of pod and DC, the ReplicationController will almost always bounce the pod. Not only does the new pod not have the initContainer, but it also gets a new name due to the use of generateName (this last part is different from the statefulset situation).

**Comment by sseago** [comment]
As for workarounds, while I haven't tried this in the StatefulSet context, for the DeploymentConfig case, there were two workarounds we've used that worked. The one that is probably most applicable to the StatefulSet situation would be to do a two-phase restore. First restore with statefulset added to excludeResources on the restore -- this will allow the pods to restore and run restic without the owning StatefulSet bouncing the pods. Then once restore is complete, do a second restore. You could just restore the StatefulSets at this point, using include-resources instead of exclude, or you could just restore everything and Velero will discard those items that already exist.

**Comment by Lyndon-Li** [comment]
I tried to reproduce the problem locally, but I couldn't reproduce it.

I also checked Kubernetes statefulset controller code, I didn't see any behavior that the controller will kill the pod if the pod's templates don't match to the original one, or it means, adding a init-container to the pod won't cause the pod to be killed by statefulset container.

Therefore, I suspect this is not a generic problem.
